File: python/second/detect_squares.py
from collections import defaultdict
class DetectSquares(object):

    def __init__(self):
        self.freq = defaultdict(int)
        self.points = []
        

    # O(1) time | O(n) space
    def add(self, point):
        """
        :type point: List[int]
        :rtype: None
        """
        self.freq[(point[0], point[1])] += 1
        self.points.append(point)
        

    # O(n) time | O(n) space
    def count(self, point):
        """
        :type point: List[int]
        :rtype: int
        """
        res = 0
        x = point[0]
        y = point[1]
        for pt in self.points:
            px = pt[0]
            py = pt[1]

            if abs(x - px) != abs(y - py) or x == px or y == py:
                continue

            res += self.freq[(x, py)] * self.freq[(px,y)]
        return res
        
solution = DetectSquares()

solution.add([3, 10])
solution.add([11, 2])
solution.add([3, 2])
solution.add([11, 2])
print(solution.count([11, 10]))

File: src/main/java/july_2025/DetectSquares.java
package july_2025;

import java.util.*;

public class DetectSquares {

    // Intuition:
    // - how to detect square? len of side is the same
    // - how to detect squares coords (x,y) - on diagonal
    // Approach:
    // - add all points to List<int[]>
    // - add freq of points
    // - loop through to count res
    // Complexity:
    // - O(n) time | O(n) space
    // Code:

    Map<List<Integer>, Integer> pointsFreq;
    List<int[]> points;

    public DetectSquares() {
        pointsFreq = new HashMap<>();
        points = new ArrayList<>();
    }

    public void add(int[] point) {
        List<Integer> key = Arrays.asList(point[0], point[1]);
        pointsFreq.put(key, pointsFreq.getOrDefault(key, 0) + 1);
        points.add(point);
    }

    public int count(int[] point) {
        int res = 0;
        int x = point[0];
        int y = point[1];

        for (int[] pt : points) {
            int px = pt[0];
            int py = pt[1];

            if (Math.abs(x - px) != Math.abs(y - py) || px == x || py == y) {
                continue;
            }

            res += pointsFreq.getOrDefault(Arrays.asList(x, py), 0)
                    * pointsFreq.getOrDefault(Arrays.asList(px, y), 0);
        }
        return res;
    }

}
